base/array.jl

```diff
@@ -2345,7 +2345,7 @@ function filter!(f, a::AbstractVector)
 
     for acurr in a
         if f(acurr)
-            a[i] = acurr
+            @inbounds a[i] = acurr
```

Comment:
Is there a policy for `@inbounds`, `AbstractArray` and the iteration protocol. I.e is it ok to segfault if someone implemented the iteration protocol erroneously?

Reply:
This is iterating over the index space of a: `foreach(i -> a[i] = a[i], eachindex(a))`. I think it's OK to assume that is required to hold.

Reply:
AFAIK the policy is that `AbstractArray` implementations which opt-in to using `@boundscheck` assert that they are correct, so it's OK to crash if that's not the case. See https://github.com/JuliaLang/julia/issues/20469 and linked issues.